### PR TITLE
Initial port of PathAliasDelete with test. Currently works for deleting ...

### PR DESCRIPTION
**Pull request by stephenpurkiss**
...a single url alias.

The original D7 function was the create or delete URL alias and thus had two inputs - source and alias. If the source was missing but an alias provided it would delete the alias, and if the alias was missing and a source provided then it would delete aliases associated with that source. 

Now that the create and delete functions are split, the delete only needs one input, i.e. what you want to delete. This initial patch provides the code to delete a single url alias as the class name suggests. 

I believe further discussion is needed to decide how this functionality is going to work in 8, i.e. can we still pass it a source and delete all relevant aliases (how do we check whether it's source or alias?) - can we send it an array of sources and/or aliases, etc. 

So, for the moment, this is a start.

### CODE REVIEW

File: src/Plugin/Action/PathAliasDelete.php [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public function execute() {
+    $language = $this->getContextValue('language');
+    $langcode = isset($language) ? $language->getId() : LanguageInterface::LANGCODE_NOT_SPECIFIED;
+
+    $source_or_alias = $this->getContextValue('source_or_alias');
+
+    $this->aliasStorage->delete($source_or_alias, $langcode);
```
Comment (klausi) [inline src/Plugin/Action/PathAliasDelete.php]: I think this is wrong - AliasStorageInterface tells me that the delete() method expects a $conditions array?

File: src/Plugin/Action/PathAliasDelete.php [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Provides a 'Delete any path alias or aliases for path' action.
+ *
+ * @Action(
+ *   id = "rules_path_alias_delete",
+ *   label = @Translation("Delete any path alias or aliases for path"),
+ *   context = {
+ *     "source_or_alias" = @ContextDefinition("string",
```
Comment (klausi) [inline src/Plugin/Action/PathAliasDelete.php]: Mixing source and alias in one parameter? Why?

File: src/Plugin/Action/PathAliasDelete.php (hunk 1 above)
Comment (stephenpurkiss) [inline src/Plugin/Action/PathAliasDelete.php]: Probably - at this moment I'm learning D8 and don't fully understand the expectations of a $conditions array, examples scarce. Pointers welcome ;)

Comment (fago) [inline src/Plugin/Action/PathAliasDelete.php]: According to to docs this needs a conditions passed, so I do not think this can work.

Comment (stephenpurkiss) [inline src/Plugin/Action/PathAliasDelete.php]: I'm a bit lost between mock data, condition arrays, and the general mind-shift back to encapsulation so it's taking a while to figure this out - I've created a new branch (getting lost in git too;) over at https://github.com/stevepurkiss/rules/tree/pull_105_16540926_conditions_array so you can see progress and if you get chance any hints welcome ;)

File: src/Plugin/Action/PathAliasDelete.php (hunk 2 above)
Comment (stephenpurkiss) [inline src/Plugin/Action/PathAliasDelete.php]: This was when I was unsure as to the scope of this function - I've updated it to just $alias.

File: src/Plugin/Action/PathAliasDelete.php (hunk 1 above)
Comment (stephenpurkiss) [inline src/Plugin/Action/PathAliasDelete.php]: OK a good night's sleep made me realise I was getting mixed up between the string input to my class and the expected array input to the delete function on the alias storage.

File: tests/src/Action/PathAliasDeleteTest.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+   * @covers ::execute()
+   */
+  public function testActionExecution() {
+
+    // Test delete path alias
+    $alias = 'about/team';
+
+    $this->aliasStorage->expects($this->once())
+      ->method('delete')
+      ->with(array('alias' => $alias));
```
Comment (klausi) [inline tests/src/Action/PathAliasDeleteTest.php]: This should use short array syntax.

File: tests/src/Action/PathAliasDeleteTest.php [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  /**
+   * Tests the action execution when no language is specified.
+   *
+   * @covers ::execute()
+   */
+  public function testActionExecution() {
+
+    // Test delete path alias
```
Comment (klausi) [inline tests/src/Action/PathAliasDeleteTest.php]: Comment should end with a dot. Also the comment does not add any information, it is already clear from the function comment what we are testing?

File: tests/src/Action/PathAliasDeleteTest.php [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+   * Tests the summary.
+   *
+   * @covers ::summary()
+   */
+  public function testSummary() {
+    $this->assertEquals('Delete any path alias', $this->action->summary());
+  }
+
+  /**
+   * Tests the action execution when no language is specified.
```
Comment (klausi) [inline tests/src/Action/PathAliasDeleteTest.php]: With the current implementation it is not possible to specify a language anyway?

**Comment by klausi** [comment]
I think this looks mostly good and is almost ready. Could you squash your commits into one?

Then we need a follow-up to delete path aliases by source. Example: I want to delete all aliases for the path "user/1".

File: tests/src/Action/PathAliasDeleteTest.php (hunk 4 above)
Comment (stephenpurkiss) [inline tests/src/Action/PathAliasDeleteTest.php]: Good point, removed.

File: tests/src/Action/PathAliasDeleteTest.php (hunk 5 above)
Comment (stephenpurkiss) [inline tests/src/Action/PathAliasDeleteTest.php]: Removed text left over from previous different definition.

**Comment by stephenpurkiss** [comment]
Made changes, looking into squashing now...

**Comment by stephenpurkiss** [comment]
Still not getting squashing quite figured out yet but will do. In mean time what about coping with deleting an array of path aliases - or would that be a derivative?

**Comment by stephenpurkiss** [comment]
Apologies for the delay, still not getting squashing. I:
1. git rebase -i HEAD~4
2. change the lines below the top to 'f'
3. it squashes the commits
4. I'm on detached head. I push using force:

durpal:rules steve$ git push origin HEAD:origin/port-delete-any-url-alias-action-to-D8-2318653 --force
Counting objects: 1, done.
Writing objects: 100% (1/1), 249 bytes | 0 bytes/s, done.
Total 1 (delta 0), reused 0 (delta 0)
To https://github.com/stevepurkiss/rules.git
- f1f6fd6...00e0c90 HEAD -> origin/port-delete-any-url-alias-action-to-D8-2318653 (forced update)

...but nothing seems to change. It's probably (definitely) completely obvious what the problem is to someone who's done it before so posting here, am sure I'm just getting mixed up between branches etc.

Oh the joys of not learning this stuff earlier ;)

**Comment by jpstacey** [comment]
When you rebase, imagine the last argument on the command line being the trunk end of a branch, and the "marker" for your checkout (usually a branch name) is sitting on the tip of the branch. You basically ask git to replace the whole branch down to (but not including) the trunk commit, creating a new branch with some squashed commits, and moving your "marker" over to the new tip.

So start at the tip of the commits you want to squash:

```
git checkout port-delete-any-url-alias-action-to-D8-2318653
```

And now the easiest thing to do if you're not sure is to start by creating a local "rebaser" branch - branches are cheap - and sitting on that. It means that you can always delete this local branch and start again with that first line up there.

```
git checkout -b rebaser
# or, next time around, git branch -f rebaser && git checkout rebaser
```

You then want to pick the last shared point between your commits and fago's - that is, the point you're offering the pull request onto i.e. 8.x-3.x - and rebase with that as the "trunk":

```
git rebase -i 8.x-3.x
# Make sure 8.x-3.x actually exists locally; if not, check it out first
```

Change all except one commit to "squash" or "fixup", and then when you get the option, edit the commit message so it makes sense (fixup might not give you the option).

When you're done, you should still be on `rebaser`, but `rebaser` no longer points to the same commit as it used to. So although `rebaser` and `port-delete-any-url-alias-action-to-D8-2318653` were the same, they're not any longer: `rebaser` has rebased to a stubby, single-commit branch.

So. `rebaser` is basically what fago wants at this stage. But where you go from there, with an existing pull request, I'm not sure and it might be down to fago to provide input. Normally I would check everything looks good in gitk (or gitx, or [git lol](https://gist.github.com/jpstacey/962289)) and then run:

```
# Force the local branch label over from the old branch to new rebaser
git branch -f port-delete-any-url-alias-action-to-D8-2318653
# Check that branch label out
git checkout port-delete-any-url-alias-action-to-D8-2318653
# Force-push it up to github
git push origin --force
```

However, because of this existing pull request, then force-pushing the `port-delete-any-url-alias-action-to-D8-2318653` might confuse github (this pull request could be tied to that commit at the tip of the old, discarded branch; I just don't know. I'm loath to recommend a `--force` if I don't know how remote apps are going to work.)

One option is that you could call `rebaser` something more descriptive e.g. `port-delete-any-url-alias-action-to-D8-rebased-2318653` and then offer a pull request on that, superseding this one? It depends a bit on fago's preferred workflow here, maybe?

**Comment by klausi** [comment]
Github can deal just fine with rebased pull requests :)

A forced push makes the old commits disappear and only the squashed one remains.

**Comment by stephenpurkiss** [comment]
OK I was rebasing to old code so I've followed the instructions but pretty sure that's wrong, but what do I know? Fairly sure the commit shouldn't have that many changes in it, just my code. Anyways, will have a play again tomorrow. Thanks again for your patience ;)

**Comment by stephenpurkiss** [comment]
I think, perhaps, maybe, I've managed to sort this out on the latest commit. Understanding git a bit more now after a few more youtubes watched ;)

**Comment by stephenpurkiss** [comment]
hmm.. but the build failed :(

**Comment by stephenpurkiss** [comment]
Test complaining that RulesUnitTestBase not found, seem to remember RulesIntegrationTestBase being introduced so changing to that, if that passes then I get a go at rebasing again :dancers:

**Comment by stephenpurkiss** [comment]
Hm maybe I rebased wrong somewhere or maybe just too far down the rabbit branch on this issue - will await someone with more knowledge than me to advise, I'm going to look into the delete aliases for path thing next.

**Comment by stephenpurkiss** [comment]
...i.e. is it supposed to be unit or integration test?

**Comment by stephenpurkiss** [comment]
hm ignore that one

**Comment by stephenpurkiss** [comment]
ok let's try that again ;)

**Comment by stephenpurkiss** [comment]
OK the travis failed on not finding unittestbase but not for my files - did I rebase wrong or is the latest commit I've done correct? (thanks and soz for the hassle, will be better next time;)

**Comment by klausi** [comment]
I think you mixed up some commits here? Why do I see a ListAddItem action in the files changed?

And we still have 4 commits here, which should be squashed into one.

**Comment by klausi** [comment]
The problem with the missing unit test base file is resolved now, Drupal core changed and we had to move the unit tests around.

**Comment by stephenpurkiss** [comment]
Obviously messed up my local version keeping up with updates as my 8.x-3.x is reporting "This branch is 3 commits ahead of fago:8.x-3.x". Also I worked on another issue on a different branch and that's messed up in it too. Working on it ;)

**Comment by stephenpurkiss** [comment]
OK, I'm back to being "This branch is even with fago:8.x-3.x"... getting there...

**Comment by stephenpurkiss** [comment]
OMG I think that's it. Well, I'm going to suggest a change soon anyway from PathAliasDelete to AliasDelete because the next one I'm working on is to delete aliases for a path for which the most appropriate name is PathAliasesDelete which would cause confusion with PathAliasDelete, and as we are referring to only an alias on this issue we can drop the 'Path'. Anyway guess that's for once this is sorted. Unless you agree and I could do it in this pull request?

**Comment by stephenpurkiss** [comment]
Renaming this as I am now using PathAliasDelete for deleting a path alias (https://github.com/fago/rules/pull/109), so this is now AliasDelete. Also updated path to unit tests.

**Comment by stephenpurkiss** [comment]
GREEN! We like green ;)

**Comment by klausi** [comment]
Finally, thanks for your endurance :-)

**Pull request merged**